pol-core/bscript/compiler/astbuilder/ExpressionBuilder.cpp

```diff
@@ -21,12 +27,70 @@ std::unique_ptr<Expression> ExpressionBuilder::expression( EscriptParser::Expres
   location_for( *ctx ).internal_error( "unhandled expression" );
 }
 
+std::unique_ptr<FunctionCall> ExpressionBuilder::function_call(
+    EscriptParser::FunctionCallContext* ctx, const std::string& scope )
+{
+  auto method_name = text( ctx->IDENTIFIER() );
+
+  auto arguments = value_arguments( ctx->expressionList() );
+
+  auto function_call = std::make_unique<FunctionCall>( location_for( *ctx ), scope, method_name,
+                                                       std::move( arguments ) );
+
+  std::string key = scope.empty() ? method_name : ( scope + "::" + method_name );
+  workspace.function_resolver.register_function_link( key, function_call->function_link );
+
+  return function_call;
+}
+
 std::unique_ptr<Expression> ExpressionBuilder::primary( EscriptParser::PrimaryContext* ctx )
 {
   if ( auto literal = ctx->literal() )
+  {
     return value( literal );
+  }
+  else if ( auto par_expression = ctx->parExpression() )
+  {
+    return expression( par_expression->expression() );
+  }
+  else if ( auto f_call = ctx->functionCall() )
+  {
+    return function_call( f_call, "" );
+  }
+  else if ( auto scoped_f_call = ctx->scopedFunctionCall() )
+  {
+    return scoped_function_call( scoped_f_call );
+  }
 
   location_for( *ctx ).internal_error( "unhandled primary expression" );
 }
 
+std::unique_ptr<FunctionCall> ExpressionBuilder::scoped_function_call(
+    EscriptParser::ScopedFunctionCallContext* ctx )
+{
+  return function_call( ctx->functionCall(), text( ctx->IDENTIFIER() ) );
+}
+
+std::vector<std::unique_ptr<Argument>> ExpressionBuilder::value_arguments(
+    EscriptGrammar::EscriptParser::ExpressionListContext* ctx )
+{
+  std::vector<std::unique_ptr<Argument>> arguments;
+
+  if ( ctx )
+  {
+    for ( auto argument_context : ctx->expression() )
+    {
+      auto loc = location_for( *argument_context );
+
+      std::string name;
```

Comment:
`name` is here (and always blank) until we can [handle passing arguments by name](https://github.com/polserver/polserver/blob/ecompile-antlr/pol-core/bscript/compiler/astbuilder/ExpressionBuilder.cpp#L543), for which we need the BinaryOperator and Identifier AST nodes.